### PR TITLE
MSUnmerged basic skeleton.

### LINKED ISSUE #10402: Create MSUnmerged basic skeleton
**Impact of the new feature**
ReqMgr2MS

**Is your feature request related to a problem? Please describe.**
It concerns a new microservice to carry out further cleanup of the sites' unmerged area (removal of files that are no longer needed/locked in the system).

**Describe the solution you'd like**
With this issue, it's expected to:
* have a bare minimum functional microservice called MSUnmerged (without any service logic);
* create the deployment scripts;
* create the deployment frontend rules;
* create the deployment scripts in services_config gitlab repository (preprod and prod branches, maybe test as well);
* create the CMSKubernetes yaml file (under this area: https://github.com/dmwm/CMSKubernetes/tree/master/kubernetes/cmsweb/services)

The docker image seems to be unique for reqmgr2ms, so the same image is used for all microservices (at this very moment).
 
**Describe alternatives you've considered**
We need to discuss with the HTTP team/Imran about which port to use; and also to add the new frontend rules.

**Additional context**
Depends on https://github.com/dmwm/WMCore/issues/10345

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #10402 

#### Status
ready

#### Description
Provides the basic skeleton for MSUnmerged

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
 * have a bare minimum functional microservice called MSUnmerged (without any service logic) - https://github.com/dmwm/WMCore/pull/10598
 * create the deployment scripts - https://github.com/dmwm/deployment/pull/1059
 * create the deployment frontend rules - https://github.com/dmwm/deployment/pull/1060
 * create the deployment scripts in services_config gitlab repository (preprod and prod branches, maybe test as well) -   
    * https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/89 
    * https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/90
 * create the CMSKubernetes yaml file  - https://github.com/dmwm/CMSKubernetes/pull/604

### CODE REVIEW

**Comment by todor-ivanov** [comment]
Hi @amaltaro 
I have fixed the missing MSUnmerged class from my previous commit.
I think I managed to create all the configuration related PRs and all are referred in the updated issue comment. Please take a look at the your convenience. Thanks you in advance.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+"""
+File       : MSUnmerged.py
+
+Description:
+
+This MicroService is meant to remove all the files from the unmerged area of the.
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:6]: Just to avoid misinterpretations, I would suggest to rephrase it to something like '...remove all the no longer needed files ...'

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    def __init__(self, msConfig, logger=None):
+        """
+        Runs the basic setup and initialization for the MSUnmerged module
+        :param msConfig: micro service configuration
+        """
+        super(MSUnmerged, self).__init__(msConfig, logger=logger)
+
+        self.msConfig.setdefault("verbose", True)
+        self.msConfig.setdefault("interval", 60)
+        self.msConfig.setdefault("services", ['unmerged'])
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:40]: I think this `services` key is only needed at the MSManager level, right?

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        self.msConfig.setdefault('limitSitesPerCycle', 100)
+        self.msConfig.setdefault('limitSitesTier', ['T1', 'T2', 'T3'])
+        self.msConfig.setdefault("rucioAccount", "FIXME_RUCIO_ACCT")
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:43]: If this is for the consistency check communication, that will be done via a REST API and there is no need to authenticate within Rucio. So it can probably go away as well.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        """
+        Executes the whole MSUnmerged logic
+        :return: summary
+        """
+        # start threads in MSManager which should call this method
+        summary = dict(UNMERGED_REPORT)
+
+        try:
+            siteList = self.getSiteList()
+            self.updateReportDict(summary, "total_num_sites", len(requestRecords))
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:56]: You probably meant `len(siteList)`? BTW, I would suggest to not use the word `site` anywhere in this microservice. We are not dealing with sites, only with storage (RSEs).

**Review by amaltaro — CHANGES_REQUESTED** [review]
These initial changes are looking good, Todor. There are a few change that need to be applied though.
I'm going to review the other deployment scripts as well and give my feedback over there. Thanks

**Comment by amaltaro** [comment]
And just a reminder, we will have to open that scrape port for Prometheus. @muhammadimranfarooqi

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 3 above)
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:43]: As we chatted over slack, disregard this comment. I had forgotten that LFN to PFN conversion will be done through Rucio RESTful API.

File: src/python/WMCore/MicroService/MSManager.py [hunk 5]
```diff
@@ -134,6 +136,17 @@ def __init__(self, config=None, logger=None):
                                                        self.logger))
             self.logger.info("--- Running %s thread %s", thname, self.ruleCleanerThread.running())
 
+        # initialize unmerged module
+        if 'unmerged' in self.services:
+            self.msUnmerged = MSUnmerged(self.msConfig, logger=self.logger)
+            thname = 'MSUnmerged'
+            self.unmergedThread = start_new_thread(thname, daemon,
+                                                   (self.unmerged,
+                                                    reqStatus,
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSManager.py:145]: Good catch! Thanks!

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 1 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:6]: Agreed

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 2 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:40]: Haven't check. I was just following the pattern from the previous micro service minimal PR. I can remove it and see what will happen ;)

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 4 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:56]: Good catch again! fixing that one too ! Thanks!

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 1 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:6]: done

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 2 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:44]: done

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 4 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py]: Agreed and done

File: src/python/WMCore/MicroService/DataStructs/DefaultStructs.py [hunk 6]
```diff
@@ -76,3 +76,15 @@
...
                           failed_request_transition=0)
+
+# summary metrics for the MSUnmerged thread
+# (also available through the `info` REST API)
+UNMERGED_REPORT = dict(thread_id="",
+                       start_time=0,
+                       end_time=0,
+                       execution_time=0,
+                       error="",
+                       total_num_sites=0,
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/DataStructs/DefaultStructs.py]: done

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    def execute(self):
+        """
+        Executes the whole MSUnmerged logic
+        :return: summary
+        """
+        # start threads in MSManager which should call this method
+        summary = dict(UNMERGED_REPORT)
+
+        try:
+            RSEList = self.getRSEList()
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:61]: Can be dealt with in your next PR as well, but I would suggest not to start variable names with upper case (it's actually against our pylintrc defnition)

**Review by amaltaro — APPROVED** [review]
Thanks for making these changes, Todor. It looks pretty good to me.
I have made a couple of comments along the code, but I'm happy if you say you will sort them in your next PR ;)

Can you please squash these commits? I'm going to check the other PRs once again.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 7 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:61]: I knew you would say that ;) fixing it.

File: src/python/WMCore/MicroService/MSManager.py [hunk 8]
```diff
@@ -49,6 +50,16 @@ def daemon(func, reqStatus, interval, logger):
         sleep(interval)
 
 
+def daemonOpt(func, interval, logger):
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSManager.py:53]: This is a planned change. It will be worked in this issue [1], when we are about to face the need for variable arguments passed to the daemon caller anyway. 

[1]
https://github.com/dmwm/WMCore/issues/10599

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 7 above)
Comment (todor-ivanov) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py]: done

**Comment by todor-ivanov** [comment]
Thanks for the review @amaltaro. I fixed the final bit you requested and squashed those commits in a single one. Please proceed at your convenience.

**Review by amaltaro — APPROVED** [review]
Thanks for making these changes here and in the other PRs, Todor. It all looks good to me.
